File: 프로그래머스/1/12906. 같은 숫자는 싫어/같은 숫자는 싫어.py
'''
1. 아이디어
- 연속적으로 나타나는 숫자는 하나만 남기고, 전부 제거한다 (중복제거&순서유지)
- 스택에 하나씩 넣되, 이전에 넣은 수와 같은 수라면 pass
2. 시간복잡도
- O(N) 하나씩 원소를 확인하며 넣어야 한다
3. 변수
- int[]: stack 기능을 구현할 정수 리스트
'''

def solution(arr):
    stack = []
    
    for i in arr:
        if stack and stack[-1] == i:
            continue
        else:
            stack.append(i)
        
    return stack
        

File: 프로그래머스/1/12906. 같은 숫자는 싫어/같은 숫자는 싫어.java
/*
연속 수 제거 및 순서 유지 -> set 불가능
stack에 하나씩 넣으면서 peek과 동일하면 pass, 다르면 push
다 넣었으면, 꺼내서 처음부터 넣기

*/
import java.util.*;

public class Solution {
    public int[] solution(int []arr) {
        Stack<Integer> stack = new Stack<>();
        
        for(int num : arr){
            if(stack.isEmpty()){
                stack.push(num);
            }else{
                int pre = stack.peek();
                if(pre == num){
                    continue;
                }else{
                    stack.push(num);
                }
            }
        }
        
        int[] result = new int[stack.size()];
        for(int i=stack.size()-1; i>=0; i--){
            result[i] = stack.pop();
        }
        
        return result;
    }
}